docs/writing-rules/glossary.md

```diff
@@ -0,0 +1,124 @@
+
+## Finding
+
+A finding is the core result of Semgrep's analysis. Findings are generated when a Semgrep rule matches a piece of code. Findings can be security issues, bugs, or code that doesn't follow coding conventions.
+
+## Fully-qualified name
+
+A fully-qualified name refers to a name which uniquely identifies a class, method, type, or module. Languages such as C# and Ruby use `::` to distinguish between fully-qualified names and regular names.
+
+Not to be confused with **tokens**.
+
+## l-value (left-, or location-value)
+
+An expression that denotes an object in memory; a memory location, something that you can use in the left-hand side (LHS) of an assignment.
+
+Compare to [r-value](https://learn.microsoft.com/en-us/cpp/c-language/l-value-and-r-value-expressions?view=msvc-170).
+
+## Metavariable
+
+A metavariable is an abstraction that lets you match something even when you don't know exactly what it is you want to match. It is similar to capture groups in regular expressions. All metavariables begin with a `$` and can only contain uppercase characters, digits, and underscores.
+
+## Propagator
+
+A propagator is any code that alters a piece of data as the data moves across the program. This includes functions, reassignments, and so on.
+
+## Rule (Semgrep rule)
+
+A rule is a specification of the patterns that Semgrep must match to the code to generate a finding. Rules are written in YAML. Without a rule, the engine has no instructions on how to match code.
+
+Rules can be run on either Semgrep OSS Engine or Semgrep Pro Engine.
+
+There are two types of rules: search and taint.
+
+<dl>
+<dt>Search rules</dt>
+<dd>Rules default to this type. Search rules simply detect matches based on the patterns described by a rule.</dd>
+<dt>Taint rules</dt>
+<dd>Taint rules make use of Semgrep's taint analysis in addition to default search functionalities. Taint rules are able to specify sources, sinks, and propagators of data as well as sanitizers of that data. For more information, see Taint analysis documentation (tk link)</dd>
+</dl>
+
+<!-- how can we say that search rules are semantic if no analysis is performed on the value of data, such as variables? Or are there levels of semantic understanding that semgrep can perform? -->
+
+## Sanitizers
+
+A sanitizer is any function that can clean untrusted or tainted data. Data from untrusted sources, such as user inputs, may be tainted with unsafe characters. Sanitizer functions ensure that unsafe characters are removed or stripped from the input.
+
+An example of a sanitizer is the [<i class="fas fa-external-link fa-xs"></i> `DOMPurify.sanitize(dirty);`](https://github.com/cure53/DOMPurify) function from the  DOMPurify package in Javascript.
+
+## Single-file analysis
+
+Also known as intrafile analysis. This refers to a tool's ability to trace or track data and its transformations across functions, such as when a variable is defined in one function but used in another.
```

Comment:
Technically, intrafile analysis only means that Semgrep performs its analyses within a file. We do use it as a shorthand for cross-function analysis, but that's mostly out of laziness.

Reply:
Saying "cross-function" or "interprocedural" doesn't provide a boundary, though? If I were new to SAST, I might wrongly conflate cross-function with cross-file because functions can be across different files?

Reply:
I guess the question is, from the perspective of the user, what is the between interprocedural and intrafile?

Reply:
I think I would say that "interprocedural taint" and "pro intrafile analysis" are the same thing. Later, "pro intrafile analysis" might include more things, but right now its main benefit over oss analysis is interprocedural taint.

However, "intrafile analysis" is still a standard term and I'm not comfortable with us redefining it even though we do do this as a shorthand.

Reply:
So - it's not good for us to conflate interprocedural with intrafile. Got it.